jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/query/preprocessor/UserAndGroupsTasksPreprocessor.java

```diff
@@ -0,0 +1,109 @@
+import org.dashbuilder.dataset.filter.ColumnFilter;
+import org.dashbuilder.dataset.filter.CoreFunctionFilter;
+import org.dashbuilder.dataset.filter.DataSetFilter;
+import org.kie.api.task.UserGroupCallback;
+import org.kie.internal.identity.IdentityProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UserAndGroupsTasksPreprocessor extends UserTasksPreprocessor{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserAndGroupsTasksPreprocessor.class);
+
+    private IdentityProvider identityProvider;
+
+    private UserGroupCallback userGroupCallback;
+
+    private String columnId;
+
+    public UserAndGroupsTasksPreprocessor(IdentityProvider identityProvider, UserGroupCallback userGroupCallback, 
+                                          String columnId, DataSetMetadata metadata) {
+        super(metadata);
+        this.identityProvider = identityProvider;
+        this.userGroupCallback = userGroupCallback;
+        this.columnId = columnId;
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    public void preprocess(DataSetLookup lookup) {
+        if (identityProvider == null || userGroupCallback == null) {
+            return;
+        }
+
+        ColumnFilter columnFilter;
+        List<Comparable> orgEntities = new ArrayList<>();
+
+        if (lookup.getFirstFilterOp() != null) {
+            List<String> potOwners = new ArrayList<String>();
+            List<ColumnFilter> columnFilters = lookup.getFirstFilterOp().getColumnFilterList();
+            Iterator<ColumnFilter> it= columnFilters.iterator();
+            while (orgEntities.isEmpty() && it.hasNext()) {
+                CoreFunctionFilter column = (CoreFunctionFilter)it.next();
+
+                if (column.getColumnId().toUpperCase().equals(columnId)) {
+                    potOwners.addAll(column.getParameters());
+
+                    for (String potOwner : potOwners) {
+                        orgEntities.addAll(Optional.ofNullable(userGroupCallback.getGroupsForUser(potOwner)).orElse(new ArrayList<>()));
+                        orgEntities.add(potOwner);
+                    }
+
+                    it.remove();
+                }
+            }
+
+            if (orgEntities.isEmpty()) {
+                AddUserAndGroupsFromIdentityProvider(orgEntities);
+            }
+
+            columnFilter = AND(in(COLUMN_POTOWNER, orgEntities));
+            lookup.getFirstFilterOp().addFilterColumn(columnFilter);
+        } else {
+            DataSetFilter filter = new DataSetFilter();
+            AddUserAndGroupsFromIdentityProvider(orgEntities);
+
+            columnFilter = AND(in(COLUMN_POTOWNER, orgEntities));
+            filter.addFilterColumn(columnFilter);
+            lookup.addOperation(filter);
+        }
+
+        LOGGER.debug("Adding column filter: {}", columnFilter);
+
+        super.preprocess(lookup);
+    }
+
+    private void AddUserAndGroupsFromIdentityProvider(List<Comparable> orgEntities) {
```

Comment:
please use lower case as first letter of the method name

Reply:
Done